### PR TITLE
Remove obsolete CMR code and refactor existing facades

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

Old CMR code is removed. The bulk of the PR is simply deleting obsolete code.
The remainder is refactoring to account for the altered facades and tweaked data model now in use:
- deleted unused facades
- moved crossmodel facade from controller to model
- deleted OfferedApplication entity

The next step will be to remove relating to model.application. We'll require an offer to be made first.

## QA steps

Perform CMR smoke test

### CODE REVIEW

**Comment by babbageclunk** [comment]
I find the last sentence of the description really hard to parse - do you mean this?

The next step will be to remove relating to model.application. We'll require an offer to be made first.

**Comment by wallyworld** [comment]
Fixed the description.

File: state/applicationoffers.go [hunk 1]
```diff
@@ -158,32 +149,35 @@ func (s *applicationDirectory) AddOffer(offer crossmodel.ApplicationOffer) (err
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return s.makeApplicationOffer(doc)
 }
 
 // UpdateOffer replaces an existing offer at the same URL.
-func (s *applicationDirectory) UpdateOffer(offer crossmodel.ApplicationOffer) (err error) {
-	defer errors.DeferredAnnotatef(&err, "cannot update application offer %q", offer.ApplicationName)
+func (s *applicationOffers) UpdateOffer(offerArgs crossmodel.AddApplicationOfferArgs) (_ *crossmodel.ApplicationOffer, err error) {
```
Comment (babbageclunk) [inline state/applicationoffers.go]: Ooh, I haven't seen a named return with the name _. I'll use that next time I need to do funny error handling stuff.

File: apiserver/crossmodel/base_test.go [hunk 2]
```diff
@@ -4,48 +4,44 @@
...
-	"github.com/juju/juju/apiserver/params"
 	"github.com/juju/juju/apiserver/testing"
 	jujucrossmodel "github.com/juju/juju/core/crossmodel"
 	jujutesting "github.com/juju/juju/juju/testing"
 )
 
 const (
-	addOffersBackendCall  = "AddOffers"
-	listOffersBackendCall = "ListOffers"
+	addOffersBackendCall  = "addOfferCall"
```
Comment (babbageclunk) [inline apiserver/crossmodel/base_test.go]: I understand this now, ignore me.

File: apiserver/application/application.go [hunk 3]
```diff
@@ -804,34 +800,49 @@ func (api *API) processRemoteApplication(url jujucrossmodel.ApplicationURL, alia
...
-	// Create a remote application entry in the model for the consumed service.
-	offer := offers.Offers[0]
-	sourceModelTag, err := names.ParseModelTag(offer.SourceModelTag)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	remoteApp, err := api.saveRemoteApplication(sourceModelTag, url.ApplicationName, url.ApplicationName, url.String(), offer.Endpoints)
-	return remoteApp, err
+	return nil, errors.NotSupportedf("non local application URL")
+	// TODO(wallyworld) - we want to support application offers instead of direct model access
```
Comment (wallyworld) [inline apiserver/application/application.go]: Keeping it as it will be resurrected next PR.

File: cmd/juju/crossmodel/crossmodel.go [hunk 4]
```diff
@@ -2,22 +2,18 @@
...
 	"github.com/juju/juju/apiserver/params"
 	"github.com/juju/juju/cmd/modelcmd"
 )
 
-var logger = loggo.GetLogger("juju.cmd.juju.crossmodel")
-
 // CrossModelCommandBase is a base structure to get cross model managing client.
 type CrossModelCommandBase struct {
-	modelcmd.ControllerCommandBase
+	modelcmd.ModelCommandBase
```
Comment (wallyworld) [inline cmd/juju/crossmodel/crossmodel.go]: Correct!

File: apiserver/crossmodel/crossmodel.go [hunk 5]
```diff
@@ -19,255 +16,171 @@ import (
...
-				errorsByURL[url] = err
-			}
-			continue
-		}
-		for _, offer := range offers.Offers {
-			foundOffers[offer.ApplicationURL] = offer
+	offers, err := api.applicationOffers.ListOffers(filters...)
+	if err != nil {
+		for _, url := range urls.ApplicationURLs {
+			errorsByURL[url] = err
```
Comment (wallyworld) [inline apiserver/crossmodel/crossmodel.go]: Yeah true, it is now deleted.Yeah, true, now deleted

**Comment by wallyworld** [comment]
$$merge$$
